Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,254 @@
+pairing, composition, primitive recursion, and μ-recursion.
+
+## References
+
+* [carneiro2019] Carneiro, Mario.
+  *Formalizing Computability Theory via Partial Recursive Functions*.
+  arXiv preprint arXiv:1810.08380, 2018.
+* [soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+
+## Tags
+
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+
+open Primrec Nat.Partrec Part
+
+/--
+The type of partial functions `f` that are recursive in an oracle `g` is the smallest type
+containing the constant zero, the successor, left and right projections, the oracle `g`,
+and is closed under pairing, composition, primitive recursion, and μ-recursion.
+
+Equivalently one can say that `f` is turing reducible to `g` when `f` is recursive in `g`.
+-/
+
+inductive RecursiveIn : (ℕ →. ℕ) → (ℕ →. ℕ) → Prop
+  | zero (g) : RecursiveIn (fun _ => 0) g
+  | succ (g) : RecursiveIn Nat.succ g
+  | left (g) : RecursiveIn (fun n => (Nat.unpair n).1) g
+  | right (g) : RecursiveIn (fun n => (Nat.unpair n).2) g
+  | oracle (g) : RecursiveIn g g
+  | pair (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => (Nat.pair <$> f n <*> h n)) g
+  | comp (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => h n >>= f) g
+  /-- primitive recursion -/
+  | prec (f h g : ℕ →. ℕ) (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y ih => do
+          let i ← ih
+          h (Nat.pair a (Nat.pair y i)))) g
+  /-- μ-recursion -/
+  | rfind (f g : ℕ →. ℕ) (hf : RecursiveIn f g) :
+      RecursiveIn (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n))) g
+
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+abbrev TuringEquivalent (f g : ℕ →. ℕ) : Prop :=
+  AntisymmRel RecursiveIn f g
+
+/--
+Custom infix notation for `TuringEquivalent`.
+-/
+infix:50 " ≡ᵀ " => TuringEquivalent
+
+/--
+If a function is partial recursive, then it is recursive in every partial function.
+-/
+lemma Nat.Partrec.recursiveIn (f : ℕ →. ℕ) (pF : Nat.Partrec f) (g : ℕ →. ℕ) :
+    RecursiveIn f g := by
+  induction pF with
+  | zero =>
+    apply RecursiveIn.zero
+  | succ =>
+    apply RecursiveIn.succ
+  | left =>
+    apply RecursiveIn.left
+  | right =>
+    apply RecursiveIn.right
+  | pair _ _ ih1 ih2 =>
+    apply RecursiveIn.pair
+    · apply ih1
+    · apply ih2
+  | comp _ _ ih1 ih2 =>
+    apply RecursiveIn.comp
+    · apply ih1
+    · apply ih2
+  | prec _ _ ih1 ih2 =>
+    apply RecursiveIn.prec
+    · apply ih1
+    · apply ih2
+  | rfind _ ih =>
+    apply RecursiveIn.rfind; apply ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma RecursiveIn.partrec_of_zero (f : ℕ →. ℕ) (fRecInZero : RecursiveIn f fun _ => Part.some 0) :
```

Comment:
This could be more generally `RecursiveIn.partrec_of_partrec (f g : ℕ →. ℕ) (hg : Nat.Partrec g) (fRecInZero : RecursiveIn f g) : Nat.Partrec f` (or, if [this](https://github.com/leanprover-community/mathlib4/pull/20301/files#r1899026178) suggestion is implemented this would just be a corollary of a (suitable form of) `RecursiveIn.trans`)